devices/src/virtio/vsock/unix/muxer.rs

```diff
@@ -0,0 +1,736 @@
+            }
+
+            // Data is ready to be read from a host-initiated connection. That would be the "connect"
+            // command that we're expecting.
+            Some(EpollListener::LocalStream(_)) => {
+                if let Some(EpollListener::LocalStream(mut stream)) = self.remove_listener(fd) {
+                    Self::read_local_stream_port(&mut stream)
+                        .and_then(|peer_port| Ok((self.allocate_local_port(), peer_port)))
+                        .and_then(|(local_port, peer_port)| {
+                            self.add_connection(
+                                ConnMapKey {
+                                    local_port,
+                                    peer_port,
+                                },
+                                MuxerConnection::new_local_init(
+                                    stream,
+                                    uapi::VSOCK_HOST_CID,
+                                    self.cid,
+                                    local_port,
+                                    peer_port,
+                                ),
+                            )
+                        })
+                        .unwrap_or_else(|err| {
+                            info!("vsock: error adding local-init connection: {:?}", err);
+                        })
+                }
+            }
+
+            _ => {
+                info!("vsock: unexpected event: fd={:?}, evset={:?}", fd, evset);
+            }
+        }
+    }
+
+    /// Parse a host "connect" command, and extract the destination vsock port.
+    ///
+    fn read_local_stream_port(stream: &mut UnixStream) -> Result<u32> {
+        let mut buf = [0u8; 32];
+        stream.read(&mut buf).map_err(Error::UnixRead)?;
+
+        let mut word_iter = std::str::from_utf8(&buf)
+            .map_err(|_| Error::InvalidPortRequest)?
+            .split_whitespace();
+
+        word_iter
+            .next()
+            .ok_or(Error::InvalidPortRequest)
+            .and_then(|word| {
+                if word.to_lowercase() == "connect" {
+                    Ok(())
+                } else {
+                    Err(Error::InvalidPortRequest)
+                }
+            })
+            .and_then(|_| word_iter.next().ok_or(Error::InvalidPortRequest))
+            .and_then(|word| word.parse::<u32>().map_err(|_| Error::InvalidPortRequest))
+            .map_err(|_| Error::InvalidPortRequest)
+    }
+
+    /// Add a new connection to the active connection pool.
+    ///
+    fn add_connection(&mut self, key: ConnMapKey, conn: MuxerConnection) -> Result<()> {
+        // We might need to make room for this new connection, so let's sweep the kill queue first.
+        // It's fine to do this here because:
+        // - unless the kill queue is out of sync, this is a pretty inexpensive operation; and
+        // - we are under no pressure to respect any accurate timing for connection termination.
+        self.sweep_killq();
```

Comment:
Just wondering, what happens when there are no connections set to expire, and the map is full? Do we just stop adding new connections? It looks like the expiry timer is set on a connection in some very specific situations, and it does not trigger on pure idleness after getting established. Does this translate to the user (or the guest) having to close some of the connections explicitly before more can be added?

Reply:
Not sure I completely understand. By "idleness," do you mean valid, established connections that have had no data flow for a while? If that's correct, then yes, we keep those as valid from the vsock protocol perspective. IMO, detecting idleness and disconnecting would be the job of a higher-level protocol. I think...